processing/src/main/java/io/druid/segment/filter/InFilter.java

```diff
@@ -85,24 +86,48 @@ public boolean apply(String input)
   @Override
   public ValueMatcher makeMatcher(ValueMatcherFactory factory)
   {
-    return factory.makeValueMatcher(
-        dimension, new Predicate<String>()
-        {
-          @Override
-          public boolean apply(String input)
-          {
-            if (extractionFn != null) {
-              input = extractionFn.apply(input);
-            }
-            return values.contains(Strings.nullToEmpty(input));
-          }
-        }
-    );
+    return factory.makeValueMatcher(dimension, getPredicate());
   }
 
   @Override
   public boolean supportsBitmapIndex(BitmapIndexSelector selector)
   {
     return selector.getBitmapIndex(dimension) != null;
   }
+
+  private DruidCompositePredicate getPredicate()
+  {
+    if (extractionFn == null) {
+      return new DruidCompositePredicate()
+      {
+        @Override
+        public boolean apply(Object inputObj)
+        {
+          String input = inputObj == null ? null : inputObj.toString();
+          return values.contains(Strings.nullToEmpty(input));
+        }
+
+        @Override
+        public boolean applyLong(long value)
+        {
+          return values.contains(String.valueOf(value));
```

Comment:
Like Selector, this should convert `values` to longs and then compare them directly against `long value`.